src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -183,6 +122,153 @@ public GeneratorDriver AddAdditionalTexts(ImmutableArray<AdditionalText> additio
             return FromState(newState);
         }
 
+        public GeneratorDriverRunResult GetRunResult()
+        {
+            var trees = ArrayBuilder<SyntaxTree>.GetInstance();
+            var diagnostics = ArrayBuilder<Diagnostic>.GetInstance();
```

Comment:
Do we need to compute these up front or just do it if/when the first time they're asked for? Not an urgent change needed here but my gut here is the IDE will only ever be walking the Results collection directly, so this might be allocations we're not using. That said, they're pretty cheap. :smile:

Reply:
Yeah, its an interesting trade off. If we don't compute up front we have to have the results be a class so we can lazy generate them. Which is probably fine. Equally I don't know that we *need* them? It seemed like a nice API for a user, but kinda ultimately designing blind. 

I see a couple options: 
- leave as-is. 
- Make it a class and make them lazy. 
- Add counts instead of the actual objects (so it's still easy to say 'does this have any trees or diagnostics')
- Remove the struct altogether and just return a collection of `GeneratorRunResult`. We can always add extension helper methods to get the data if people request it.

---
In reply to: [465383851](https://github.com/dotnet/roslyn/pull/46296#discussion_r465383851) [](ancestors = 465383851)

Reply:
FWIW i think i'm actually leaning towards the last option.

---
In reply to: [466089088](https://github.com/dotnet/roslyn/pull/46296#discussion_r466089088) [](ancestors = 466089088,465383851)